Docker compose for local environments

chillu opened the pull request:
Mostly a copy from https://github.com/silverstripeltd/cwp-demo-template/blob/master/docker-compose.yml,
but with the ability to mount /var/www/html as the local development folder (rather than just copy contents as a one-off)

This isn't required to *build* the docker environment for usage in the Kubernetes cluster,
but a more accurate way to develop locally (compared to Vagrant etc). Note that it contains Solr even though we're not using it yet. It's easy to remove later if we decide to go with ElasticSearch, or upgrade if we decide to use Solr8.

See https://github.com/silverstripeltd/open-sourcerers/issues/107

chillu committed:
Docker compose for local environments

Mostly a copy from https://github.com/silverstripeltd/cwp-demo-template/blob/master/docker-compose.yml,
but with the ability to mount /var/www/html as the local development folder (rather than just copy contents as a one-off)

This isn't required to *build* the docker environment for usage in the Kubernetes cluster,
but a more accurate way to develop locally (compared to Vagrant etc)

See https://github.com/silverstripeltd/open-sourcerers/issues/107

dnsl48 commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,47 @@
+    build: .
+    restart: always
+    ports:
+      - 8200:80
```
why not "80:80"? changing ports usually makes it more confusing

chillu replied:
Because 80 is often taken on the host environment already - most devs here run a webserver on the host permanently. I'm just using the defaults from the CWP docker compose file here.

dnsl48 commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,47 @@
+    image: solr4
+    restart: always
+    ports:
+      - 8983:8983
```
the best practice is to take ports into quotes (like `- "8983:8983"`), otherwise YAML may sometimes interpret them as a single sexagesimal number. (not this one though)

chillu replied:
Hah yeah I remember reading about this, I think in a Puppet context? Good catch, fixed the two `ports` assignments in the file now.

dnsl48 commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,47 @@
+
+services:
+  solr:
+    image: solr4
```
```
Pulling solr (solr4:)...
ERROR: pull access denied for solr4, repository does not exist or may require 'docker login'
```

dnsl48 replied:
I guess we may need to pull https://github.com/silverstripeltd/cwp-demo-template/tree/master/solr-template into `docker/solr` and then have `build: ./solr` instead of `image: solr4`

chillu replied:
Damn, yeah I prebuilt this box when running `cwp-demo-template`, so it didn't complain when I took over the code for bambusa. I've changed it to build locally on the fly now.

See https://github.com/silverstripeltd/cwp-demo-template#building-a-new-image for the original steps, which were a bit more clunky.

dnsl48 commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,47 @@
+      - data:/var/www/html/.solr
+      # This overlays the folder copied in Dockerfile
+      - ..:/var/www/html
+    links:
```
`links` is [deprecated](https://docs.docker.com/compose/compose-file/#links) and not required as long as the services belong to a single network.
```
Links are not required to enable services to communicate - by default,
any service can reach any other service at that service’s name.
```

chillu replied:
Cool, removed, and verified it still builds.

dnsl48 commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,47 @@
+    environment:
+      MYSQL_ROOT_PASSWORD: password
+  web:
+    build: .
```
```
Step 3/12 : COPY ./docker/php.ini /usr/local/etc/php/conf.d/custom.ini
ERROR: Service 'web' failed to build: COPY failed: stat /var/lib/docker/tmp/docker-builder473337291/docker/php.ini: no such file or directory
```

We may need to change it to:
```
build:
      context: ..
      dockerfile: docker/Dockerfile
```

Otherwise it builds the image with the context of `/docker` rather than `/`

chillu replied:
Sweet, changed

dnsl48 commented on README.md:
````diff
@@ -9,13 +9,46 @@ This project is meant to demonstrate key features of SilverStripe CMS. It combin
+docker-compose exec web /bin/bash
+``` 
+
+Note that you should run commands as the web user, through `gosu www-data`.
````
could do `docker-compose exec -u www-data web /bin/bash` instead

chillu replied:
Hah I tried that, but didn't work. I think I did `docker exec` rather than `docker-compose exec`.

chillu committed:
Quoted ports in YAML

chillu committed:
Describe how to docker build solr

chillu committed:
Removed deprecated "links" from docker compose

chillu committed:
Changed docker compose root

It has trouble with paths in referenced bash scripts
otherwise (e.g. copying php.ini)

chillu committed:
Document easier way to get accurate docker user

chillu commented:
@dnsl48 Thanks for the in-depth review! Ready for your re-review now

dnsl48 reviewed (approved):
looks good

chillu commented:
OK, the build is broken for the same reason as `master`: https://travis-ci.org/silverstripe/bambusa-installer/builds/553470191. Serge has approved, so I'm merging.
